src/stratigraphy/groundwater/groundwater_extraction.py

```diff
@@ -0,0 +1,420 @@
+
+        Args:
+            extracted_date (str): The extracted date.
+
+        Returns:
+            bool: True if the date is correct, otherwise False.
+        """
+        return self.compare_dates(self.date, extracted_date)
+
+    def is_extracted_depth_correct(self, extracted_depth: float) -> bool:
+        """Checks if the extracted depth is correct.
+
+        Args:
+            extracted_depth (float): The extracted depth.
+
+        Returns:
+            bool: True if the depth is correct, otherwise False.
+        """
+        return self.depth == extracted_depth
+
+    def is_extracted_elevation_correct(self, extracted_elevation: float) -> bool:
+        """Checks if the extracted elevation is correct.
+
+        Args:
+            extracted_elevation (float): The extracted elevation.
+
+        Returns:
+            bool: True if the elevation is correct, otherwise False.
+        """
+        return self.elevation == extracted_elevation
+
+
+class GroundwaterLevelExtractor:
+    """Extracts coordinates from a PDF document."""
+
+    def __init__(self, document: fitz.Document):
+        """Initializes the CoordinateExtractor object.
+
+        Args:
+            document (fitz.Document): A PDF document.
+        """
+        self.doc = document
+        self.groundwater_keys = read_params("matching_params.yml")["groundwater_keys"]
+
+    def find_groundwater_key(self, lines: list[TextLine], allowed_errors: int = 3) -> list[TextLine] | None:  # noqa: E501
+        """Finds the location of a groundwater key in a string of text.
+
+        Args:
+            lines (list[TextLine]): Arbitrary text lines to search in.
+            allowed_errors (int, optional): The maximum number of errors (Levenshtein distance) to consider a key
+                                            contained in text. Defaults to 3 (guestimation; no optimisation done yet).
+
+        Returns:
+            TextLine | None: The line of the drilling method key found in the text.
+        """
+        matches = []
+        for key in self.groundwater_keys:
+            if len(key) < 5:
+                # if the key is very short, do an exact match
+                pattern = regex.compile(r"\b" + key + r"\b", flags=regex.IGNORECASE)
+            else:
+                pattern = regex.compile(r"\b" + key + "{e<" + str(allowed_errors) + r"}\b", flags=regex.IGNORECASE)
+
+            for line in lines:
+                match = pattern.search(line.text)
+                if match:
+                    matches.append((line, sum(match.fuzzy_counts)))
+
+        # if no match was found, return None
+        if len(matches) == 0:
+            return None
+
+        # Remove duplicates
+        matches = list(dict.fromkeys(matches))
+
+        # Sort the matches by their error counts (ascending order)
+        matches.sort(key=lambda x: x[1])
+
+        # Return the top three matches (lines only)
+        return [match[0] for match in matches[:5]]
```

Comment:
"top three" and "matches[:5]" don't seem to fit together?
What's the motivation for only considering an limited number of matches anyway?

Reply:
Thank you for catching this. I started with the three best matches but realized that the performance was better with five. I think here the logic of limiting the number of matches was to improve the speed (not a big performance boost, but still) as well as to avoid using any match that could reduce the performance of extraction. Therefore, using the five best seemed like a good compromise. As you can see, the list of groundwater keywords is quite long, as there are many different representations in the different documents. This is a bit tricky to handle and the more words we have the higher is the chance for a false positive.